### PR TITLE
Enable RDRAND for i586, too

### PR DESCRIPTION
**Pull request by mtremer**
IPFire is being compiled for i586 omitting some instructions for i686. However, RDRAND is available on some systems and can of course be used.

### CODE REVIEW

**Comment by abelbeck** [comment]
Hi @mtremer , out of curiosity which i586 CPU's have the RDRAND instruction ?

I seem to recall it takes 22nm lithography to get the necessary quantum effects.

Though when cross-compiling a person may have a generic i586 32-bit toolchain running on a i686 CPU.

**Comment by abelbeck** [comment]
@nhorman possibly a more general configure `--without-rdrand` and `--with-rdrand` could override the auto-build-check since at runtime you test for `rdrand` anyway.

**Comment by nhorman** [comment]
@abelbeck  I still fail to see why anyone would want to to remove rdrand at compile time, as it has a runtime check for the rdseed/rdrand instructions, as well as a runtime administrative disable flag.  The only advantage I can see is the code space savings.  If you want to save that 36kb we can talk, but I'd like to discuss the use case that calls for that minutiae of space savings, and is willing to sacrifice the entropy source to get it.

@mtremer I'm not opposed to this change, but as @abelbeck says, I'm not aware of any i586 system that supports the RDRAND instruction.  If you can point out what processor it is that requires this change, we can document and merge it

**Comment by mtremer** [comment]
Hello altogether,

this indeed not obvious why I need this change.

Back in time, we ran some benchmarks what code optimisations perform
best on some hardware and what is most compatible (the common
denominator was the AMD Geode LX at that time) which almost supports
the full i686 instruction set, but not popl. So we downgraded to i586
which runs perfectly on all other systems and even is often faster than
i686 because of smaller size and small caches on platforms like the
Intel Atom.

Long story short, this is still our default instruction set that is
very compatible but some systems have RDRAND which can be used no
matter what.

I would agree with compiling it is on x86 no matter what, but probably
this just wastes space on ARM, PPC and so on...

**Comment by mtremer** [comment]
Sorry for causing confusion. So we are compiling out toolchain for
i586-pc-linux-gnu and we use these CFLAGS:

  -march=i586 -mtune=generic -fomit-frame-pointer

That compiled rng-tools well for every version so far and RDRAND used
to work before the version 6 release.

**Comment by abelbeck** [comment]
I just tried this PR's patch on our i586 toolchain and it understood the `rdrand` instruction just fine. (But `rdseed` needs to be poked (.byte) in as rng-tools v5 did.)
```
$ /home/dev/astlinux/trunk/output/host/usr/bin/i586-unknown-linux-gnu-gcc --version
i586-unknown-linux-gnu-gcc (crosstool-NG 1.20.0) 4.8.3
```
rng-tools build snippet...
```
/home/dev/astlinux/trunk/output/host/usr/bin/i586-unknown-linux-gnu-gcc -DHAVE_CONFIG_H -I.  -I./jitterentropy-library -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64   -pthread -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -MT rngd-rngd_rdrand.o -MD -MP -MF .deps/rngd-rngd_rdrand.Tpo -c -o rngd-rngd_rdrand.o `test -f 'rngd_rdrand.c' || echo './'`rngd_rdrand.c
mv -f .deps/rngd-rngd_rdrand.Tpo .deps/rngd-rngd_rdrand.Po
/home/dev/astlinux/trunk/output/host/usr/bin/i586-unknown-linux-gnu-gcc -DHAVE_CONFIG_H -I.  -I./jitterentropy-library -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64  -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -MT rdrand_asm.o -MD -MP -MF .deps/rdrand_asm.Tpo -c -o rdrand_asm.o rdrand_asm.S
mv -f .deps/rdrand_asm.Tpo .deps/rdrand_asm.Po
/home/dev/astlinux/trunk/output/host/usr/bin/i586-unknown-linux-gnu-gcc -DHAVE_CONFIG_H -I.  -I./jitterentropy-library -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64   -pthread -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -MT rngd-rngd_jitter.o -MD -MP -MF .deps/rngd-rngd_jitter.Tpo -c -o rngd-rngd_jitter.o `test -f 'rngd_jitter.c' || echo './'`rngd_jitter.c
mv -f .deps/rngd-rngd_jitter.Tpo .deps/rngd-rngd_jitter.Po
/home/dev/astlinux/trunk/output/host/usr/bin/i586-unknown-linux-gnu-gcc   -pthread -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -pthread  -o rngd rngd-rngd.o rngd-rngd_entsource.o rngd-rngd_linux.o rngd-util.o  rngd-rngd_rdrand.o rdrand_asm.o  rngd-rngd_jitter.o  librngd.a -lsysfs ./jitterentropy-library/libjitterentropy.a     -lsysfs
```

**Comment by nhorman** [comment]
Sorry, that still doesn't wash with me.  I586 was released in 1993, and has absolutely no clue about the rdrand and/or rdseed instructions.  @abelbeck I don't know how you built your compiler, but we already established that you had to enable some extra isa sets to get rdrand recognized as an instruction from the previous issue you opened (though I don't see them enabled in that snippet, so something in your environment has changed).

@mtremer I don't know whats working for you, or how it might have worked prior to version 6, but Its just not possible.  Please show me a true i586/pentium 1 cpu that actually supports the rdrand instruction.  If gcc is accepting the rdrand mnemonic for i586 targets, then surely you must see that gcc has a bug there.  Imagine if we failed to do a runtime cpuid check for rdrand, and built targeting i586 but got an rdrand instruction in there.  It would immediately machine check

**Comment by abelbeck** [comment]
>  @abelbeck I don't know how you built your compiler, but we already established that you had to enable some extra isa sets to get rdrand recognized as an instruction from the previous issue you opened (though I don't see them enabled in that snippet, so something in your environment has changed).

Recap: Originally `rdrand` was not compiling for me, I tried a lot of things, finally this patch made it work for us: (no changes to the toolchain or CFLAGS)
```
--- rng-tools-6.3.1/rdrand_asm.S.orig	2018-08-29 09:52:46.773547700 -0500
+++ rng-tools-6.3.1/rdrand_asm.S	2018-08-29 09:55:17.549915865 -0500
@@ -58,7 +58,7 @@
 1:
 	mov	$RDRAND_RETRY_LIMIT, %r10d
 2:
-	rdseed	%rax
+	.byte	0x48,0x0f,0xc7,0xf8	/* rdseed %rax */
 	jnc	3f
 	mov	%rax, (%rdi)
 	add	$8, %rdi
@@ -140,7 +140,7 @@
 1:
 	mov	$RDRAND_RETRY_LIMIT, %ecx
 2:
-	rdseed	%eax
+	.byte	0x0f,0xc7,0xf8		/* rdseed %eax */
 	jnc	3f
 	mov	%eax, (%edi)
 	add	$4, %edi
```
@mtremer FYI rng-tools v5 included `.byte` workarounds for both `rdseed` and `rdrand` so the compiler did not need to understand them.
Ref: `rdrand_asm.S` here https://sourceforge.net/projects/gkernel/files/rng-tools/5/

@nhorman Please do what you determine is best, I understand you need to look forward and perform some cleanups.

**Comment by mtremer** [comment]
I applied this patch to our build tree and let it compile over night. It compiled and it has rdrand and rdseed instructions in the binary:

```[ms@rice-oxley ~]$ file mnt/usr/sbin/rngd 
mnt/usr/sbin/rngd: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux.so.2, for GNU/Linux 3.2.0, stripped
[ms@rice-oxley ~]$ objdump -D mnt/usr/sbin/rngd | grep -E "rd(rand|seed)"
 804dad2:	0f c7 f2             	rdrand %edx
 804db1b:	0f c7 f8             	rdseed %eax
 804db39:	0f c7 f0             	rdrand %eax```

It is this build here: https://nightly.ipfire.org/next/2018-09-09%2018:27:56%20+0100-a0d612be/i586/ and the build log with rng-tools in it is here: https://nightly.ipfire.org/next/2018-09-09%2018:27:56%20+0100-a0d612be/i586/log/_build.ipfire.log

**Comment by abelbeck** [comment]
I also performed some testing ...

Tests performed on a Jetway Mini-ITX board
```
Intel(R) Celeron(R) CPU  N2930  @ 1.83GHz
```
In both cases, only enable RDRAND w/o AES using rngd args: `-x0 -x1 -x5 -O 2:use_aes:0`

1) 32-bit, i586 compatible gcc toolchain, using this PR's patch so rng-tools's `rdrand` is enabled
```
pbx3 ~ # uname -a
Linux pbx3 3.16.57-astlinux #1 SMP PREEMPT Mon Sep 10 06:10:31 CDT 2018 i686 GNU/Linux

pbx3 ~ # cat /proc/sys/kernel/random/entropy_avail
3107

pbx3 ~ # cat /dev/random | rngtest -c1000
rngtest 6.4
Copyright (c) 2004 by Henrique de Moraes Holschuh
This is free software; see the source for copying conditions.  There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

rngtest: starting FIPS tests...
rngtest: bits received from input: 20000032
rngtest: FIPS 140-2 successes: 999
rngtest: FIPS 140-2 failures: 1
rngtest: FIPS 140-2(2001-10-10) Monobit: 0
rngtest: FIPS 140-2(2001-10-10) Poker: 0
rngtest: FIPS 140-2(2001-10-10) Runs: 1
rngtest: FIPS 140-2(2001-10-10) Long run: 0
rngtest: FIPS 140-2(2001-10-10) Continuous run: 0
rngtest: input channel speed: (min=2.369; avg=2.693; max=3.155)Mibits/s
rngtest: FIPS tests speed: (min=66.227; avg=71.376; max=72.523)Mibits/s
rngtest: Program run time: 7352503 microseconds
```

2) 64-bit, x86_64 gcc toolchain
```
pbx3 ~ # uname -a
Linux pbx3 3.16.57-astlinux #1 SMP PREEMPT Fri Sep 7 16:21:32 CDT 2018 x86_64 GNU/Linux

pbx3 ~ # cat /proc/sys/kernel/random/entropy_avail
3100

pbx3 ~ # cat /dev/random | rngtest -c1000
rngtest 6.4
Copyright (c) 2004 by Henrique de Moraes Holschuh
This is free software; see the source for copying conditions.  There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

rngtest: starting FIPS tests...
rngtest: bits received from input: 20000032
rngtest: FIPS 140-2 successes: 1000
rngtest: FIPS 140-2 failures: 0
rngtest: FIPS 140-2(2001-10-10) Monobit: 0
rngtest: FIPS 140-2(2001-10-10) Poker: 0
rngtest: FIPS 140-2(2001-10-10) Runs: 0
rngtest: FIPS 140-2(2001-10-10) Long run: 0
rngtest: FIPS 140-2(2001-10-10) Continuous run: 0
rngtest: input channel speed: (min=3.357; avg=3.910; max=5.289)Mibits/s
rngtest: FIPS tests speed: (min=19.583; avg=63.519; max=75.990)Mibits/s
rngtest: Program run time: 5178972 microseconds
```

Other than the slower 32-bit vs. 64-bit `rdrand` reads, rng-tools works fine with a compiler backward compatible to i586.

We will be including this PR's patch in our project if @nhorman does not choose to accept it here.

**Comment by abelbeck** [comment]
@mtremer Thanks for the patch.

**Comment by mtremer** [comment]
No problem :)

**Comment by nhorman** [comment]
but....it is the architecture that you are optimizing for.  You said exactly that here:
https://github.com/nhorman/rng-tools/pull/31#issuecomment-419749531

You are building for an i586 system, which doesn't support rdrand instructions.  The fact that you are running on more modern hardware that does support it is irrelevant, because the compiler doesn't know what hardware you are running on.  All it knows is that you said you would be running on i586 hardware which means it can only emit i586 instructions.  As such, the fact that it emits rdrand instructions is a bug.  If you actually ran this code on a true i586 cpu, it would take a machine check the instant it tried to get entropy from the rdrand source, and thats a problem.

**Comment by abelbeck** [comment]
>  If you actually ran this code on a true i586 cpu, it would take a machine check the instant it tried to get entropy from the rdrand source, and thats a problem.

No, since `rdrand` is tested for at runtime it is not a problem, as per my test on a i586 CPU:
https://github.com/nhorman/rng-tools/pull/31#issuecomment-419910131

It would seem compilers are tolerant to pass on special-case instructions that must be tested at runtime like `rdrand` and `aes-ni`.

The only problem with this PR is if the compiler failed to handle `rdrand` and there is no method to configure `--without-rdrand` to work around the build failure.

**Comment by nhorman** [comment]
@abelbeck ok, if you want to drill into the details, sure, it will still run because we test for rdrand at runtime, but as I noted before that is so completely besides the point.  If you tell your compiler to emit code for a given arch, it doesn't grant exceptions just because you tried to emit a specific mnemonic.   It just doesn't work that way.  You can see it in the gcc documentation.  The -march flag documents which arches support which ISA extensions, and i586 clearly does not include rdrnd/rdseed.  The fact that it gcc allows the emission of that instruction for that arch is a bug, and no matter how much that happens to fill a need for you use case, it doesn't make it ok for a compiler to emit instructions that aren't part of the isa.  If that were the case then gcc would just skip all the complex arch filtering code alltogether and have programmers use intrinsics to lay down AVX/MMX/SSE*/etc instructions whenever you like.
